### PR TITLE
Implement a debug_enqueue! macro that writes to a ring buffer flushed only upon panicking.

### LINKED ISSUE #1652: [RFC] Panic buffer
While debugging https://github.com/tock/tock/issues/1651, I've found it useful to be able to write to some ring buffer that is only printed in the panic handler. Together with a panic button (https://github.com/tock/tock/issues/1640), this makes it much easier to debug components related to the regular debugging itself (e.g. Segger RTT which relies on alarms).

### Proposal

Adding a `debug_panic!` macro to `kernel/src/debug.rs`, quite similar to `debug!` in terms of usage, except that instead of writing to a (virtual) UART immediately, it just appends the bytes to a ring buffer. However, contrary to `debug!`, new bytes continue to be appended when the buffer is full, so that the contents are always the most recent N bytes.

Then, inside the panic handler, the current contents of this ring buffer are printed in the diagnostics (so, the last N bytes are shown in the panic diagnostic).

### Configuration

Given that such a buffer would be useful in specific cases (in general one can use the regular `debug!`), I'd suggest to have the size of this buffer configurable in `kernel/src/config.rs` to avoid overhead in the general case. A buffer size of 0 would completely disable it.

### PR DESCRIPTION
**Pull request by gendx**
### Pull Request Overview

This pull request fixes #1652.

### Testing Strategy

This pull request was tested on an nRF52840-DK board, writing some data with `debug_enqueue!` and triggering a panic to observe the result in the console:
- by adding a `DebugQueue` component in the board,
- without setting the `DebugQueue` buffer.

Overflowing the ring buffer and using formatting was also tested:

```rust
let buffer = static_init!([u8; 128], [0; 128]);
DebugQueueComponent::new(buffer).finalize(());

for i in 0..1000 {
    debug_enqueue!("foo({})", i);
}
```

gives the following output.

```
---| Flushing debug queue:
(987)
foo(988)
foo(989)
foo(990)
foo(991)
foo(992)
foo(993)
foo(994)
foo(995)
foo(996)
foo(997)
foo(998)
foo(999)
```

### TODO or Help Wanted

N/A

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

### CODE REVIEW

**Comment by ppannuto** [comment]
I like this a lot, but I wonder if we could benefit from making it even a touch more generic. The real benefit I think is to provide a mechanism to keep debug printing from affecting timing (too much). I don't think that's useful only in panic situations.

Maybe this could be thought of as more of a generic debug queue and support methods like:
 - `debug_enqueue()`
 - `debug_flush()`

With the `debug_flush()` being called automatically on `panic!`.

In this way, we could wire up flush events to other interesting moments (button press, SPI transaction complete) without requiring a panic to get the benefit of the queuing behavior.

**Comment by gendx** [comment]
That could also be useful.

If I understood correctly, `debug_flush` would then pass the bytes to the writer used by `debug`, is that correct?

**Comment by gendx** [comment]
Before bikeshedding on the names, I added a `debug_panic_flush` macro that does what @ppannuto suggested.

File: kernel/src/debug.rs [hunk 1]
```diff
@@ -40,7 +40,6 @@
 use core::cell::Cell;
```
Comment (bradjc) [inline kernel/src/debug.rs]: Add example in "Example" block.

**Comment by ppannuto** [comment]
bors d+

I don't think this needs much of `last-call` window, it's been open for a little while. I'd say go ahead and merge tomorrow.

**Pull request merged**